src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,903 @@
+  double const y_ratio = static_cast<double>((src.n_rows - 1)) / height;
+  for(size_t row = 0; row != dst.n_rows; ++row)
+  {
+    size_t y = static_cast<size_t>(row * y_ratio);
+    double const y_diff = (row * y_ratio) - y; //distance of the nearest pixel(y axis)
+    double const y_diff_2 = 1 - y_diff;
+    for(size_t col = 0; col != dst.n_cols; ++col)
+    {
+      size_t x = static_cast<size_t>(col * x_ratio);
+      double const x_diff = (col * x_ratio) - x; //distance of the nearet pixel(x axis)
+      double const x_diff_2 = 1 - x_diff;
+      double const y2_cross_x2 = y_diff_2 * x_diff_2;
+      double const y2_cross_x = y_diff_2 * x_diff;
+      double const y_cross_x2 = y_diff * x_diff_2;
+      double const y_cross_x = y_diff * x_diff;
+      dst(row, col) = y2_cross_x2 * src(y, x) +
+              y2_cross_x * src(y, x + 1) +
+              y_cross_x2 * src(y + 1, x) +
+              y_cross_x * src(y + 1, x + 1);
+    }
+  }
+
+  return dst;
+}
+
+template<typename MatType, typename CubeType>
+CubeType StructuredForests<MatType, CubeType>::
+sepFilter2D(CubeType& InImage, arma::vec& kernel, int radius)
+{
+  CubeType OutImage = this->CopyMakeBorder(InImage, radius, radius, radius, radius);
+
+  arma::vec row_res(1), col_res(1);
+  // reverse InImage and OutImage to avoid making an extra matrix.  
+  for(size_t k = 0; k < OutImage.n_slices; ++k)
+  {
+    for(size_t j = radius; j < OutImage.n_cols - radius; ++j)
+    {
+      for(size_t i = radius; i < OutImage.n_rows - radius; ++i)
+      {
+        row_res = OutImage.slice(k).row(i).subvec(j - radius, j + radius) * kernel;
+        col_res = OutImage.slice(k).col(i).subvec(i - radius, i + radius).t() * kernel;
+        // divide by 2: avg of row_res and col_res, divide by 3: avg over 3 locations.
+        InImage(i - radius, j - radius, k) = (row_res(0) + col_res(0)) / 2 / 3;
+      }
+    }
+  }
+
+  return InImage;
+}
+
+template<typename MatType, typename CubeType>
+CubeType StructuredForests<MatType, CubeType>::
+ConvTriangle(CubeType& InImage, int radius)
+{
+  if (radius == 0)
+  {
+    return InImage;
+  }
+  else if (radius <= 1)
+  {
+    double p = 12.0 / radius / (radius + 2) - 2;
+    arma::vec kernel = {1 , p, 1};
+    kernel = kernel / (p + 2);
+    
+    return this->sepFilter2D(InImage, kernel, radius);
+  }
+  else
+  {
+    int len = 2 * radius + 1;
+    arma::vec kernel(len);
+    for( size_t i = 0; i < radius; ++i)
+      kernel(i) = i + 1;
+    
+    kernel(radius) = radius + 1;
+    
+    for( size_t i = radius + 1; i < len; ++i)
+      kernel(i) = i - 1;
+    return this->sepFilter2D(InImage, kernel, radius);
+  }
+}
+
+//just a helper function, can't use it for anything else
+//finds max numbers on cube axis and returns max values,
+// also stores the locations of max values in Location
+template<typename MatType, typename CubeType>
+MatType StructuredForests<MatType, CubeType>::
+MaxAndLoc(CubeType& mag, arma::umat& Location)
```

Comment:
For const correctness
`MaxAndLoc(const CubeType &mag, arma::umat& Location) const`